### PR TITLE
Remove rgb::Image::ignore_alpha for simplicity

### PR DESCRIPTION
**Pull request by y-guyon**
rgb::Image::format can be set to an opaque format instead.

Also do not premultiply samples with alpha before dropping alpha. This is not a behavior I would expect.

### CODE REVIEW

**Comment by vigneshvg** [comment]
I agree that this is a welcome change. I added this just to replicate libavif's behavior.

We would have to fix the C API. I think we cannot remove it from the C API since it is intended to be compatible with libavif. So we can probably make this a no-op.

**Comment by y-guyon** [comment]
> We would have to fix the C API. I think we cannot remove it from the C API since it is intended to be compatible with libavif. So we can probably make this a no-op.

I tried to convert "ignore alpha" to "remove alpha channel" in `reformat.rs`. What do you think?

File: src/reformat/rgb.rs [hunk 1]
```diff
@@ -305,6 +296,101 @@ impl Image {
...
+                }
+                None => None,
+            },
+            row_bytes: self.row_bytes,
+        }
+    }
+
+    pub fn shuffle_channels_to(&self, format: Format) -> AvifResult<Image> {
+        if self.format == format {
+            return Ok(self.view());
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:325]: it's not immediately obvious why this view is needed. can't we just return the input image?

File: src/capi/reformat.rs [hunk 2]
```diff
@@ -54,19 +54,36 @@ impl From<rgb::Image> for avifRGBImage {
...
+            (rgb::Format::Argb, true) => rgb::Format::Rgb,
+            (rgb::Format::Argb, false) => rgb::Format::Argb,
+            (rgb::Format::Bgr, _) => rgb::Format::Bgr,
+            (rgb::Format::Bgra, true) => rgb::Format::Bgr,
+            (rgb::Format::Bgra, false) => rgb::Format::Bgra,
+            (rgb::Format::Abgr, true) => rgb::Format::Bgr,
+            (rgb::Format::Abgr, false) => rgb::Format::Abgr,
+            (rgb::Format::Rgb565, _) => rgb::Format::Rgb565,
+        };
+        if format == rgb.format {
```
Comment (vigneshvg) [inline src/capi/reformat.rs:83]: since this is the only place `shuffle_channels_to` is called, you can change the argument type of `shuffle_channels_to` to `self` instead of `&self` (i.e._) that function will consume the input image and produce a new output image without creating any copies.

you can then replace this if condition with an unconditional call to `dst.shuffle_channels_to(format).unwrap()`

File: src/reformat/rgb.rs [hunk 3]
```diff
@@ -457,4 +543,38 @@ mod tests {
         }
         Ok(())
     }
+
+    #[test]
+    fn channel_shuffling() {
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:548]: nit: you can probably use the same function name `shuffle_channels_to`.

File: src/reformat/rgb.rs [hunk 4]
```diff
@@ -457,4 +543,38 @@ mod tests {
...
+            format: Format::Rgba,
+            pixels: Some(Pixels::Buffer(vec![0u8, 1, 2, 3])),
+            row_bytes: 4,
+            ..Default::default()
+        };
+
+        let shuffled = image.shuffle_channels_to(Format::Rgba);
+        match &shuffled.unwrap().pixels {
+            Some(Pixels::Pointer(ptr)) => assert_eq!(
+                unsafe { std::slice::from_raw_parts(*ptr, 4) },
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:562]: you can use `shuffled.row(..)` to get rid of this unsafe block.

ditto below. we only care about the actual pixel values in this test and now how they are stored underneath (pointer, buffer or buffer16 does not really matter).

**Review by vigneshvg — CHANGES_REQUESTED** [review]

**Comment by vigneshvg** [comment]
> > We would have to fix the C API. I think we cannot remove it from the C API since it is intended to be compatible with libavif. So we can probably make this a no-op.
> 
> I tried to convert "ignore alpha" to "remove alpha channel" in `reformat.rs`. What do you think?

yes this approach is good.

File: src/reformat/rgb.rs (hunk 1 above)
Comment (y-guyon) [reply src/reformat/rgb.rs:325]: Not sure how to implement that.

Also there is little point into calling this function with the same format. Returning a shallow copy seemed as good as returning an `AvifError`, so I can do the latter if you prefer.

File: src/capi/reformat.rs (hunk 2 above)
Comment (y-guyon) [reply src/capi/reformat.rs:83]: > since this is the only place `shuffle_channels_to` is called

It is also used in the test. Would you advise refactoring the test or the function and calling site?

File: src/reformat/rgb.rs (hunk 1 above)
Comment (vigneshvg) [reply src/reformat/rgb.rs:325]: if you change the input parameter type from `&self` to `self` and do `return self` here it will be a "move" which i think is the Rust way to move things rather than making an explicit shallow copy.

File: src/capi/reformat.rs (hunk 2 above)
Comment (vigneshvg) [reply src/capi/reformat.rs:83]: yeah i think we should refactor the test. please see my suggestion about parameterizing the tests.

File: src/reformat/rgb.rs [hunk 5]
```diff
@@ -457,4 +543,28 @@ mod tests {
...
+            width: 1,
+            height: 1,
+            depth: 8,
+            format: Format::Rgba,
+            pixels: Some(Pixels::Buffer(vec![0u8, 1, 2, 3])),
+            row_bytes: 4,
+            ..Default::default()
+        };
+
+        for (format, expected) in [
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:559]: please make this a test parameter instead of a for loop. so that we get all the benefits (each test will run in parallel, know exactly which one failed, etc).

something like this:

```rust
use test_case::test_case;
#[test_case(Format::Rgba, &[0, 1, 2, 3])]
#[test_case(Format::Abgr, &[3, 2, 1, 0])]
#[test_case(Format::Rgb, &[0, 1, 2])]
fn shuffle_channels_to(format: Format, expected: &[u8]) {
  let image = ...;
  assert_eq!(...); // use format and expected here.
}
```

untested code. but there should be other examples in the repo that uses the test_case for generating parameterized tests.

**Review by vigneshvg — CHANGES_REQUESTED** [review]

File: src/reformat/rgb.rs [hunk 6]
```diff
@@ -319,6 +310,101 @@ impl Image {
...
+
+    pub fn shuffle_channels_to(self, format: Format) -> AvifResult<Image> {
+        if self.format == format {
+            return Ok(self);
+        }
+        if self.format == Format::Rgb565 || format == Format::Rgb565 {
+            return Err(AvifError::NotImplemented);
+        }
+
+        let mut dst = Image {
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:345]: sorry i missed this earlier.

you can use the struct update syntax for this initialization (i.e.) override only the fields that you want and use `..self` in the end.

```rust
let mut dst = Image {
  format,
  pixels: None,
  row_bytes: 0,
  ..self
};
```

there might be other places where we could have done this (like line 57), but we can leave that for a separate PR.

File: src/reformat/rgb.rs [hunk 7]
```diff
@@ -319,6 +310,101 @@ impl Image {
         }
         Ok(())
     }
+
+    pub fn view(&self) -> Image {
```
Comment (vigneshvg) [inline src/reformat/rgb.rs:314]: this function is unused now.

**Review by vigneshvg — CHANGES_REQUESTED** [review]

**Review by vigneshvg — APPROVED** [review]
